File: Python/6-LinkedList/Problems/3-LinkedListCycle/LinkedListCycle.py
class ListNode:
    def __init__(self, x):
        """
        Represents a node in a singly linked list.
        :param x: The value stored in the node.
        """
        self.val = x
        self.next = None

def create_linked_list(nodes):
    """
    Creates a linked list from a list of values.

    Args:
        nodes (list): A list of values for the nodes.

    Returns:
        ListNode: The head of the linked list.
    """
    if not nodes:
        return None
    head = ListNode(nodes[0])
    current = head
    for i in range(1, len(nodes)):
        current.next = ListNode(nodes[i])
        current = current.next
    return head

def create_cyclic_linked_list(nodes, pos):
    """
    Creates a linked list with a cycle.

    Args:
        nodes (list): A list of values for the nodes.
        pos (int): The index of the node where the cycle starts (0-indexed).

    Returns:
        ListNode: The head of the linked list with a cycle.  Returns None if pos is invalid.
    """
    if not nodes:
        return None
    head = ListNode(nodes[0])
    current = head
    cycle_start = None
    for i in range(1, len(nodes)):
        current.next = ListNode(nodes[i])
        current = current.next
        if i == pos:
            cycle_start = current
    if pos != -1 and cycle_start is not None:
        current.next = cycle_start  # Create the cycle
    return head

def display_linked_list(head, max_nodes=20):
    """
    Displays the values of a linked list, stopping at a maximum number of nodes to avoid infinite loops.

    Args:
        head (ListNode): The head of the linked list.
        max_nodes (int, optional): The maximum number of nodes to display. Defaults to 20.
    """
    current = head
    count = 0
    nodes = []
    while current and count < max_nodes:
        nodes.append(str(current.val))
        current = current.next
        count += 1
        if current == head:  # Detect a full cycle back to the head
            nodes.append("...")
            break
    if current and count >= max_nodes:
        nodes.append("...")  # Indicate list was truncated
    print(" -> ".join(nodes))

def has_cycle_approach1(head):
    """
    Approach 1: Using a set to store visited nodes.

    Time Complexity: O(n), where n is the number of nodes in the linked list.
    Space Complexity: O(n), for the set.
    """
    if not head:
        return False
    visited = set()
    current = head
    while current:
        if current in visited:
            return True
        visited.add(current)
        current = current.next
    return False

def has_cycle_approach2(head):
    """
    Approach 2: Floyd's Cycle Detection Algorithm (Tortoise and Hare).
    This is the most efficient and commonly used approach.

    Time Complexity: O(n), where n is the number of nodes in the linked list.
    Space Complexity: O(1), constant space.
    """
    if not head or not head.next:
        return False
    slow = head
    fast = head.next
    while fast and fast.next:
        if slow == fast:
            return True
        slow = slow.next
        fast = fast.next.next
    return False

def has_cycle_approach3(head):
    """
    Approach 3: Modifying the linked list (adding a visited flag).
    This approach modifies the original linked list, which might not be desirable.

    Time Complexity: O(n)
    Space Complexity: O(1)
    """
    if not head:
        return False

    current = head
    while current:
        if hasattr(current, 'visited'):
            return True
        else:
            current.visited = True #Add an attribute
            current = current.next
    return False

def has_cycle_approach4(head):
    """
     Approach 4: Using a dictionary to store node pointers and their counts.
     Detects a cycle if any node is visited more than once.

     Time Complexity: O(n)
     Space Complexity: O(n)
    """
    if not head:
        return False

    node_dict = {}
    current = head
    while current:
        if current in node_dict:
            node_dict[current] += 1
            if node_dict[current] > 1:
                return True
        else:
            node_dict[current] = 1
        current = current.next
    return False

def has_cycle_approach5(head):
    """
    Approach 5: Using a long and short pointer with a counter.  This approach is similar to Approach 2,
    but adds a counter to limit the number of steps, preventing an infinite loop in some cases
    if the cycle is incorrectly formed.

    Time Complexity: O(n) in a correct linked list.  Could be higher if the cycle is broken.
    Space Complexity: O(1)
    """
    if not head or not head.next:
        return False

    short_ptr = head
    long_ptr = head.next
    counter = 0
    max_count = 10000 # Set a maximum limit
    while long_ptr and long_ptr.next and counter < max_count:
        if short_ptr == long_ptr:
            return True
        short_ptr = short_ptr.next
        long_ptr = long_ptr.next.next
        counter += 1
    return False

if __name__ == "__main__":
    # Test cases
    test_cases = [
        ([], -1),
        ([1], -1),
        ([1, 2, 3, 4, 5], -1),
        ([1, 2, 3, 4, 5], 2),
        ([1, 2], 0),
        ([1, 2], 1),
        ([1,2,3,4,5,6,7,8,9,10], 5), # Longer list
    ]

    for nodes, cycle_pos in test_cases:
        print(f"\nTest Case: List = {nodes}, Cycle Start Position = {cycle_pos}")
        head = create_cyclic_linked_list(nodes, cycle_pos)
        display_linked_list(head)

        if cycle_pos == -1:
            print("  List has no cycle.")
        else:
            print("  List has a cycle.")

        # Test each approach
        print(f"  Approach 1 (Set):             {'Cycle' if has_cycle_approach1(head) else 'No Cycle'}")
        print(f"  Approach 2 (Floyd's):         {'Cycle' if has_cycle_approach2(head) else 'No Cycle'}")
        print(f"  Approach 3 (Modify List):     {'Cycle' if has_cycle_approach3(head) else 'No Cycle'}")
        print(f"  Approach 4 (Dictionary):      {'Cycle' if has_cycle_approach4(head) else 'No Cycle'}")
        print(f"  Approach 5 (Counter):         {'Cycle' if has_cycle_approach5(head) else 'No Cycle'}")

        # Display the list again, in case an approach modified it.
        head = create_cyclic_linked_list(nodes, cycle_pos) #Recreate
        display_linked_list(head)

File: Java/6-LinkedList/Problems/3-LinkedListCycle/LinkedListCycle.cpp
// Definition of a linked list node (needed for all solutions)
class ListNode {
  constructor(val) {
    this.val = val;
    this.next = null;
  }
}

// Helper function to create a linked list with an optional cycle
function createLinkedList(values, cyclePos = -1) {
  const head = new ListNode(values[0]);
  let current = head;
  let cycleNode = null;

  for (let i = 1; i < values.length; i++) {
    current.next = new ListNode(values[i]);
    current = current.next;
    if (i === cyclePos) {
      cycleNode = current;
    }
  }

  if (cyclePos !== -1) {
    current.next = cycleNode; // Create the cycle
  }
  return head;
}

// Helper function to visualize the linked list (with cycle or without)
function visualizeLinkedList(head, message = "") {
    let current = head;
    let nodes = [];
    let visited = new Set();
    let cycleStart = null;
    let i = 0;

    while (current && !visited.has(current) && i < 20) { // Limit to 20 to avoid infinite loops in console if cycle is present
        nodes.push(current.val);
        visited.add(current);
        current = current.next;
        i++;
        if (visited.has(current)) {
            cycleStart = current.val;
            break;
        }
    }
    let resultString = "";
    if (cycleStart)
        resultString = `${message} List: ${nodes.join(" -> ")} -> ${cycleStart} (Cycle Detected!)`;
    else
        resultString = `${message} List: ${nodes.join(" -> ")} (No Cycle)`;

    console.log(resultString);
}

// 1. Floyd's Cycle Detection (Fast and Slow Pointer)
function hasCycleFloyd(head) {
  if (!head || !head.next) {
    return false;
  }

  let slow = head;
  let fast = head.next;

  while (slow && fast && fast.next) {
    if (slow === fast) {
      return true;
    }
    slow = slow.next;
    fast = fast.next.next;
  }
  return false;
}

// 2. Using a Set to store visited nodes
function hasCycleSet(head) {
  const visited = new Set();
  let current = head;

  while (current) {
    if (visited.has(current)) {
      return true;
    }
    visited.add(current);
    current = current.next;
  }
  return false;
}

// 3.  Modifying the linked list (adding a property - not recommended for real-world use, as it changes the original list)
function hasCycleModify(head) {
  let current = head;
  while (current) {
    if (current.visited) {
      return true;
    }
    current.visited = true;
    current = current.next;
  }
  return false;
}

// 4. Using an array to store visited nodes (less efficient than Set)
function hasCycleArray(head) {
    const visitedNodes = [];
    let current = head;

    while (current) {
        if (visitedNodes.includes(current)) {
            return true;
        }
        visitedNodes.push(current);
        current = current.next;
    }
    return false;
}

// 5. Recursive approach (not ideal for long lists due to stack overflow potential)
function hasCycleRecursive(head, visited = new Set()) {
  if (!head) {
    return false;
  }

  if (visited.has(head)) {
    return true;
  }

  visited.add(head);
  return hasCycleRecursive(head.next, visited);
}



// Test cases
const list1 = createLinkedList([1, 2, 3, 4, 5]);
const list2 = createLinkedList([1, 2, 3, 4, 5], 2); // Cycle starts at node with value 3
const list3 = createLinkedList([1]);
const list4 = createLinkedList([1,2]);
const list5 = createLinkedList([1,2], 0);

visualizeLinkedList(list1, "List 1");
visualizeLinkedList(list2, "List 2");
visualizeLinkedList(list3, "List 3");
visualizeLinkedList(list4, "List 4");
visualizeLinkedList(list5, "List 5");


console.log("--- Floyd's Cycle Detection ---");
console.log("List 1 has cycle:", hasCycleFloyd(list1)); // false
console.log("List 2 has cycle:", hasCycleFloyd(list2)); // true
console.log("List 3 has cycle:", hasCycleFloyd(list3)); // false
console.log("List 4 has cycle:", hasCycleFloyd(list4)); // false
console.log("List 5 has cycle:", hasCycleFloyd(list5)); // true


console.log("--- Set Approach ---");
console.log("List 1 has cycle:", hasCycleSet(list1)); // false
console.log("List 2 has cycle:", hasCycleSet(list2)); // true
console.log("List 3 has cycle:", hasCycleSet(list3)); // false
console.log("List 4 has cycle:", hasCycleSet(list4)); // false
console.log("List 5 has cycle:", hasCycleSet(list5)); // true

console.log("--- Modification Approach ---");
console.log("List 1 has cycle:", hasCycleModify(list1)); // false
console.log("List 2 has cycle:", hasCycleModify(list2)); // true - WARNING: Modifies List
console.log("List 3 has cycle:", hasCycleModify(list3)); // false
console.log("List 4 has cycle:", hasCycleModify(list4)); // false
console.log("List 5 has cycle:", hasCycleModify(list5)); // true - WARNING: Modifies List

console.log("--- Array Approach ---");
console.log("List 1 has cycle:", hasCycleArray(list1)); // false
console.log("List 2 has cycle:", hasCycleArray(list2)); // true
console.log("List 3 has cycle:", hasCycleArray(list3)); // false
console.log("List 4 has cycle:", hasCycleArray(list4)); // false
console.log("List 5 has cycle:", hasCycleArray(list5)); // true

console.log("--- Recursive Approach ---");
console.log("List 1 has cycle:", hasCycleRecursive(list1)); // false
console.log("List 2 has cycle:", hasCycleRecursive(list2)); // true
console.log("List 3 has cycle:", hasCycleRecursive(list3)); // false
console.log("List 4 has cycle:", hasCycleRecursive(list4)); // false
console.log("List 5 has cycle:", hasCycleRecursive(list5)); // true
